# Metatranscriptomic classification in the study of microbial translocation

Antonin COLAJANNI <sup>1,2</sup>, Raluca URICARU<sup>2</sup>, Rodolphe THIÉBAUT<sup>1</sup>, and Patricia THEBAULT<sup>2</sup> 

<sup>1</sup> Univ. Bordeaux, INSERM, INRIA, BPH, U1219, F-33000 Bordeaux, France 

<sup>2</sup> Univ. Bordeaux, CNRS, Bordeaux INP, LaBRI, UMR 5800, F-33400 Talence, France

Corresponding Author: antonin.colajanni@u-bordeaux.fr 

**Contact:** 

antonin.colajanni@u-bordeaux.fr


**Keywords**

Metatranscriptomic, Metagenomic, Microbial translocation, Public data, Benchmark

**Abstract** 

Microbial translocation occurs when bacteria migrate from the gut to the blood due to gut barrier alterations, potentially triggering persistent immune activation and affecting immune responses. Translocation can be studied using whole blood RNA sequencing techniques to analyze the “meta-transcriptome”: all the RNA from bacteria, viruses and fungi from the blood. However, a key challenge is reusing cohort data, which primarily consists of human sequences, to characterize the non-human meta-transcriptome. Cohort studies typically focus on human data while minimizing non-human contamination. From a translocation perspective, these non-human sequences become the focus, requiring human sequences to be filtered out, leaving a small fraction (~2%) to be analyzed. In previous work, Douek et al. used a sequence assembly-based pipeline for translocation analysis. We compared this approach with an assembly-free method and found that integrating both strategies into a “hybrid” pipeline improved classification performance in simulations. While real-data validation remains challenging, our results suggest this hybrid strategy enhances microbial translocation analysis.
